runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,110 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
+                  //     parameters can share slots with variables of other, potentially incompatible types
+                  bool objTypeUnknown;
+                  TR_OpaqueClasBlock *objClass = NULL;
+
+                  if (objTypeSig == NULL)
+                     {
+                     objTypeUnknown = true;
+                     }
+                  else
+                     {
+                     objClass = fej9->getClassFromSignature(objTypeSig, length, dest->getSymbolReference()->getOwningMethod(comp));
+                     bool isParameter = dest->getSymbolReference()->getSymbol()->isParm();
+
+                     objTypeUnknown = objClass == NULL || objClass == comp->getObjectClassPointer() ||
+                                      TR::Compiler->cls.isInterfaceClass(comp, objClass) || isParameter;
+                     }
+
+                  //need to generate code to load dataAddr in cases (2) and (3) (unless the object is known to be NULL)
+                  mayNeedDataAddrLoad = !dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');
+
+                  //pass in dest and destOffset separately in case (3) only
+                  separateDestAndOffset = mayNeedDataAddrLoad && objTypeUnknown;
+                  }
+               else
+                  {
+                  //if for some reason we cannot get the symbol reference (and thus, the type signature)
+                  //of dest, we'll need to take the conservative approach and pass dest and destOffset to
+                  //setmemoryEvaluator() separately
+                  separateDestAndOffset = true;
+                  mayNeedDataAddrLoad = true;
+                  }
+            }
+         #endif /* J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION */
+
+            TR::Node *copyMemNode;
+
+            //setting up IL trees to be passed along to setmemoryEvaluator()
+            if (separateDestAndOffset) // CASE (3): dest and destOffset passed in separately
+               {
+               //four child setup
+               copyMemNode = TR::Node::createWithSymRef(TR::arrayset, 4, 4, dest, destOffset, len, byteValue, node->getSymbolReference());
+               }
+            else // CASE (1) and (2): dest += destoffset, then pass in to evaluator
+               {
+            #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+
+               if (mayNeedDataAddrLoad) // CASE (2) only
+                  {
+                  //load dataAddr and use as object base address (previously dest)
+                  TR::Node *dataAddrNode = J9::TransformUtil::generateDataAddrLoadTrees(comp, dest);
+                  dest = dataAddrNode;
+                  }
+
+            #endif /* J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION */
+
+               //three child setup
+               dest = TR::Node::create(comp->target().is64Bit() ? TR::aladd : TR::aiadd, 2, dest, destOffset); //dest += destOffset
```

Comment:
how does the tree look previously (prior to this work) for 32bit?
isn't `destOffset `a long type? such that you need a `l2i ` of it before it can be chained into `aiadd`, otherwise i think IL is invalid.